android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/MainActivity.java

```diff
@@ -589,11 +684,35 @@ public void onKeyboardDownloadFinished(HashMap<String, String> keyboardInfo, int
           KMManager.setKeyboard(packageID, keyboardID, languageID, keyboardName, languageName, kFont, kOskFont);
         }
       }
+    } else if (result == 0) {
+      Toast.makeText(this, "Keyboard already exists", Toast.LENGTH_SHORT).show();
     } else {
       Toast.makeText(this, "Keyboard download failed", Toast.LENGTH_SHORT).show();
```

Comment:
Can we give a clearer error message -- e.g. `"Keyboard 'keyboardID' download failed because <blah>"`?

Reply:
I can include the `keyboardID` in the toast message. 

But in order to explain why a download failed, we'd need to spec out a list of error codes (`result` < 0) that get passed to the `onKeyboardDownloadFinished()` event handler.

Reply:
OK, we can just put the error number into the message for now. That at least gives them something to report to us other than "it didn't work" :)